jangaroo-maven/jangaroo-maven-plugin/src/main/java/net/jangaroo/jooc/mvnplugin/converter/RootPackageJson.java

```diff
@@ -68,18 +63,25 @@ public PackageJson readPackageJson() {
     return this.packageJson;
   }
 
-  public void writePackageJson() {
-    try {
-      if (this.packageJson.getWorkspaces() != null) {
-        this.packageJson.getWorkspaces().sort(Comparator.naturalOrder());
-      }
-      File packageJsonFile = Paths.get(path + "/package.json").toFile();
-      if (packageJsonFile.exists()) {
-        packageJsonFile.delete();
+  public void writePackageJson(List<String> newWorkspaces) throws IOException {
+    boolean tryWriting = true;
+    File packageJsonFile = Paths.get(path + "/package.json").toFile();
+    if (!packageJsonFile.exists()) {
+      packageJsonFile.createNewFile();
+    }
+    while (tryWriting) {
+      try (RandomAccessFile randomAccessFile = new RandomAccessFile(packageJsonFile, "rw"); FileLock fileLock = randomAccessFile.getChannel().lock()) {
+        readPackageJson(randomAccessFile);
+        newWorkspaces.forEach(this.packageJson::addWorkspace);
+        if (this.packageJson.getWorkspaces() != null) {
+          this.packageJson.getWorkspaces().sort(Comparator.naturalOrder());
+        }
+        randomAccessFile.write(objectMapper.writerWithDefaultPrettyPrinter().writeValueAsString(packageJson).concat("\n").getBytes());
+        fileLock.release();
+        tryWriting = false;
+      } catch (OverlappingFileLockException e) {
+        // try again if lock is currently acquired
```

Comment:
You could also move the package.json to a resource. The only generated part is the "workspaces". Maybe load the resource with the objectMapper and add the "workspaces" entry?

Reply:
Good idea, at least to remove the declaration of the static parts of the package.json in the code. It would only run the first time executed, since after that the existing package.jsons are loaded.